properties/pkg/properties.go

```diff
@@ -0,0 +1,257 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package properties
+
+import (
+	"encoding/json"
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+)
+
+const (
+	// Default delimiter used for exporting Property
+	DefaultSeparator = "="
+)
+
+type keyIndex struct {
+	key   string
+	index uint
+}
+
+type keyIndexList []keyIndex
+
+func (k keyIndexList) Len() int {
+	return len(k)
+}
+
+func (k keyIndexList) Swap(i, j int) {
+	k[i], k[j] = k[j], k[i]
+}
+
+func (k keyIndexList) Less(i, j int) bool {
+	return k[i].index < k[j].index
+}
+
+// Properties is used to store a group of Property items belong together.
+// It also supports number of operations on the stored Property items.
+type Properties struct {
+	// Map of Property objects
+	properties map[string]Property
+	// Map of Property keys and indices for bookkeeping  their original
+	// order in the Properties document.
+	keys map[string]keyIndex
+	// Index for the next Property to be added to Properties
+	nextKeyIndex uint
+	// State lock to prevent concurrent updates of Properties
+	mutex sync.RWMutex
+}
+
+// Get a Property using by its name.
+func (p *Properties) Get(key string) (Property, bool) {
+	// Acquire read lock
+	p.mutex.RLock()
+	defer p.mutex.RUnlock()
+
+	// Get property if exists
+	propByKey, found := p.properties[key]
+	if !found {
+		return Property{}, false
+	}
+	return propByKey, true
+}
+
+// Get list of Property names.
+func (p *Properties) Keys() []string {
+	// Acquire read lock
+	p.mutex.RLock()
+	defer p.mutex.RUnlock()
+
+	// Create an keyIndexList with the size of the keys map
+	keyIdxList := make(keyIndexList, 0, len(p.keys))
+
+	// Add keyIndex items from keys map to keyIndexList before sorting
+	for _, keyIdx := range p.keys {
+		keyIdxList = append(keyIdxList, keyIdx)
+	}
+
+	// Sort keys in keyIndexList by their index
+	sort.Sort(keyIdxList)
+
+	// Create a string slice with the size of the keys map
+	// holding the properties keys in order
+	keys := make([]string, 0, len(p.keys))
+
+	// Retrieve the keys from the sorted keyIndexList
+	for _, keyIdx := range keyIdxList {
+		keys = append(keys, keyIdx.key)
+	}
+
+	return keys
+}
+
+// Get number of items in Properties
+func (p *Properties) Len() int {
+	// Acquire read lock
+	p.mutex.RLock()
+	defer p.mutex.RUnlock()
+	return len(p.keys)
+}
+
+// Add a new Property to Properties with the given name and value.
+func (p *Properties) Set(key string, value interface{}, comment string) error {
+	prop := Property{}
+	err := prop.set(key, value, comment)
+	if err != nil {
+		return err
+	}
+
+	// Acquire RW lock before updating internal state
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	p.put(prop)
```

Comment:
Let's reuse existing code.
```suggestion
	p.Put(prop)
```